lib/spack/spack/cmd/checkbinaries.py

```diff
@@ -0,0 +1,247 @@
+##############################################################################
+# Copyright (c) 2013-2018, Lawrence Livermore National Security, LLC.
+# Produced at the Lawrence Livermore National Laboratory.
+#
+# This file is part of Spack.
+# Created by Todd Gamblin, tgamblin@llnl.gov, All rights reserved.
+# LLNL-CODE-647188
+#
+# For details, see https://github.com/spack/spack
+# Please also see the NOTICE and LICENSE files for our notice and the LGPL.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License (as
+# published by the Free Software Foundation) version 2.1, February 1999.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the IMPLIED WARRANTY OF
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms and
+# conditions of the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+from __future__ import print_function
+
+from six.moves.urllib.request import urlopen
+from six.moves.urllib.error import URLError
+
+import json
+import os
+import re
+import yaml
+
+import llnl.util.tty as tty
+
+import spack.cmd
+import spack.config
+from spack.error import SpackError
+from spack.spec import Spec
+from spack.util.executable import which
+
+
+description = "check if binaries need to be rebuilt"
+section = "developer"
+level = "long"
+
+
+_buildcache_path = os.path.join('build_cache')
+_pkg_match_regex = re.compile(r'packages[/\\]([^/\\]+)[/\\]package\.py')
+_pkg_yaml_match_regex = re.compile(r'<a href=[^>]+\>\s*([^<]+)\s*\<')
+
+
+def changed_package_files(args):
```

Comment:
This should be moved to some common place like `spack.util.git` so that this command and `spack flake8` can use the same code.

Reply:
I agree with you here, if we keep this logic I'll move it somewhere it can be shared by both.  However, just recently I've been thinking that using git to determine what has changed is flawed in general.  Maybe you can straighten me out though.  If we only look at changed packages, then will we not also need, for each changed package, to somehow walk back up the dependency tree and add all dependent packages to the rebuild list?  I've recently been wondering if we just need to do a `full_hash` on every package in the release `spec-set`, since that will truly discover if any package needs to be rebuilt.  To avoid a lot of network traffic, we could think about keeping a slightly more detailed index on the binary mirror so we can fetch all the full hashes in a single request.

Reply:
> If we only look at changed packages, then will we not also need, for each changed package, to somehow walk back up the dependency tree and add all dependent packages to the rebuild list?

Yes, that's true.  Look at `spack dependents` for some code that does that -- it constructs an inverted dependency `dict` and uses that to get the static dependency information, which will tell you which packages *could* have changed.

I think computing the full hash for the full spec set is the best way to go.  We need to do something equivalent to that.  It may still be a win to run the code from `spack dependents` to figure out which packages could have changed, then only re-hash the things from the spec set that have those names, but it's probably only a small win at this point.

Reply:
Modifying the index to have more info is definitely a good idea.